axes/handlers/database.py

```diff
@@ -0,0 +1,283 @@
+    is_client_method_whitelisted,
+    is_client_username_whitelisted,
+)
+
+
+log = getLogger(settings.AXES_LOGGER)
+
+
+class AxesDatabaseHandler(AxesBaseHandler):  # pylint: disable=too-many-locals
+    """
+    Signal handler implementation that records user login attempts to database and locks users out if necessary.
+    """
+
+    def is_allowed(self, request: HttpRequest, credentials: Optional[Dict[str, Any]] = None) -> bool:
+        """
+        Check if the request or given credentials are already locked by Axes.
+
+        This function is called from
+
+        - function decorators defined in ``axes.decorators``,
+        - authentication backends defined in ``axes.backends``, and
+        - signal handlers defined in ``axes.handlers``.
+
+        This function checks the following facts for a given request:
+
+        1. Is the request IP address _blacklisted_? If it is, return ``False``.
+        2. Is the request IP address _whitelisted_? If it is, return ``True``.
+        4. Is the request HTTP method _whitelisted_? If it is, return ``True``.
+        3. Is the request user _whitelisted_? If it is, return ``True``.
+        5. Is failed authentication attempt always allowed to proceed? If it is, return ``True``.
+        6. Is failed authentication attempt count over the attempt limit? If it is, return ``False``.
+
+        Refer to the function source code for the exact implementation.
+        """
+
+        if is_client_ip_address_blacklisted(request):
+            return False
+
+        if is_client_ip_address_whitelisted(request):
+            return True
+
+        if is_client_method_whitelisted(request):
+            return True
+
+        if is_client_username_whitelisted(request, credentials):
+            return True
+
+        if not settings.AXES_LOCK_OUT_AT_FAILURE:
+            return True
+
+        # Check failure statistics against cache
+        cache_hash_key = get_cache_key(request, credentials)
+        num_failures_cached = get_axes_cache().get(cache_hash_key)
+
+        # Do not hit the database if we have an answer in the cache
+        if num_failures_cached is not None:
+            return num_failures_cached < settings.AXES_FAILURE_LIMIT
+
+        # Check failure statistics against database
+        attempts = get_user_attempts(request, credentials)
+
+        lockouts = attempts.filter(
+            failures_since_start__gte=settings.AXES_FAILURE_LIMIT,
+        )
+
+        return not lockouts.exists()
+
+    def user_login_failed(self, sender, credentials, request, **kwargs):  # pylint: disable=too-many-locals
+        """
+        When user login fails, save AccessAttempt record in database and lock user out if necessary.
+
+        :raises AxesSignalPermissionDenied: if user should is locked out
+        """
+
+        if request is None:
+            log.warning('AXES: AxesDatabaseHandler.user_login_failed does not function without a request.')
+            return
```

Comment:
@aleksihakli out of curiosity. I'm trying to remember but I can't;  why are not we returning an exception for this cases?

Reply:
This is the previous behaviour and I opted to go with backwards compatible fashion. Raising would probably be the correct way to behave here, though, I agree on that.

Reply:
cc @camilonova any thought on that?